### DISCUSSION TITLE
v41.0-beta1

### DISCUSSION THREAD

**Discussion by westnordost** [streetcomplete/StreetComplete#3802] [Release Discussions]
<h3>Measure with the camera (AR)</h3>
  <p>Measuring with AR is only supported <a href="https://developers.google.com/ar/devices#google_play_devices">on certain devices</a> and you need to have the Google Play Services on your phone. If AR is not available, the new quests are disabled by default because you'd need to use a tape measure instead. (#3709, #879)</p>
  <h3>New Quests</h3>
  <ul>
  <li><i>What’s the width of this cycleway here?</i> (#698)</li>
  <li><i>What’s the width of this road here?</i></li>
  <li><i>What is the height limit here?</i></li>
  </ul>
  <h3>Quest enhancements</h3>
  <ul>
  <li>Add "handlebar holders" as bicycle parking type option (#3061, #3724), by @sams96</li>
  <li>Tapping on road in street name sometimes did not work (#3797)</li>
  <li>Other small improvements (#3740, #3755, #3793, #3783), thanks @matkoniecz, @tiptoptom</li>
  </ul>

<hr /><em>This discussion was created from the release <a href='https://github.com/streetcomplete/StreetComplete/releases/tag/v41.0-beta1'>v41.0-beta1</a>.</em>

**Comment by tiptoptom** [comment]
A very cool feature. Unfortunately, my device is not supported. :( 

What about a function that makes it easy to measure the road width using your own feet, by placing foot to foot and using a once stored value to calculate the actual width?

**Reply by westnordost** [reply]
This won't be implemented, but it is possible to enter the distance manually. So you just need to memorize your shoe length and do basic math.

**Reply by mnalis** [reply]
If one were to write such generic app for measuring length by footsteps, perhaps that app could annonce same intents so it would be used instead of ARcore (if that is possible, I have no idea how SC-AR interaction works)?

**Comment by westnordost** [comment]
I tested the AR measurement more broadly today, measured a couple of 4-6m broad residential streets and compared it with a precise laser measurement. For me, the distance measured for 6m broad streets was reliably 10-20cm shorter than the actual width, that's the widths of the curbs or ~3%, so I think this deviation is acceptable. 
At one occasion, the measured width was completely off (4m instead of 6m) but I noticed this and measured again. I think this happens when the AR thinks the phone is closer to the ground than it really is.

Just standing on one side of the street, starting the measurement from there and then tilting the phone upwards till the arrow is on the other side of the street rarely works well on asphalt streets. What always works is to walk the measurement <strike>dog</strike> arrow over to the other side. It works so well that I jokingly thought I should replace the arrow with a 2d-sprite of a dog on a leash.
(Maybe I should display a tracking error message in situations where the arrow gets stuck that hints at that one should keep the arrow close to oneself.)

**Reply by mnalis** [reply]
Maybe add tag `source:width=ARcore` automatically if measure was taken with AR? That would help distinguish it from manually entered value.

**Comment by wielandb** [comment]
Regarding AR-Measurement:

- I noticed that when attempting to measure from one street side and just tilting the phone, the arrow will "refuse" to get a certain distance away from me, while tracking is still acurate. Was this specifically implemented or is this a quirk of AR-Core?
- I also made the experience that walking over the street with the phone is the most reliable way of getting a measurement
- If the tracking is lost midway through walking over the street and picks up again, is it garantueed to still be acurate? If not, maybe a hint could be implemented "Hey, I shortly lost tracking - make sure the start point is still where you set it"
- I noticed a bug but I havent opened an issue yet because I cannot repruduce it reliably: Sometimes when starting the activity, the camera feed will not be shown. The "movement"-indicator will still be there and even the arrow will appear in the right place and track correctly, you just can't see anything. The only consistent variable I noticed is that that only happens if the AR-measurement was used several times in that session before. Closing the activity and starting again does not fix it, while restarting the app does.
- I was wondering how much of an impact "not walking perfectly straight" across the road has, and if it would be necessary (and needed... and possible) to have some kind of mechanism that keeps the line orthogonal to the curb.
- #3809

**Reply by westnordost** [reply]
> the arrow will "refuse" to get a certain distance away from me, while tracking is still acurate. Was this specifically implemented or is this a quirk of AR-Core?

The arrow only moves on the area that has been detected by ARCore as a horizontal plane. If it is too far away, I figure, it doesn't (yet) detect it as a plane. I found no way to tell ARCore that whatever the plane on which the first point is set, just extend it to infinity (if that's possible at all).

> If the tracking is lost midway through walking over the street and picks up again, is it garantueed to still be acurate? 

I don't know, never occured for me.

> to have some kind of mechanism that keeps the line orthogonal to the curb.

I wouldn't know how

**Reply by westnordost** [reply]
> I noticed a bug but I havent opened an issue yet because I cannot repruduce it reliably: Sometimes when starting the activity, the camera feed will not be shown. The "movement"-indicator will still be there and even the arrow will appear in the right place and track correctly, you just can't see anything. The only consistent variable I noticed is that that only happens if the AR-measurement was used several times in that session before. Closing the activity and starting again does not fix it, while restarting the app does.

The ARCore session manages access to the camera and the ARCore session is closed every time the activity is closed. So, this looks like an issue in ARCore.

**Reply by westnordost** [reply]
Regarding the first point, I just implemented that as soon as the first point has been placed, the hit test with the plane will assume that the plane on which the first point is located is infinite (i.e. it is not checked if the hit is within the plane). 

- This unstucks the arrow, but not always. Apparently, ARCore always does **some** bounds check with the hit check on a plane, otherwise I'd be able to move the arrow towards the horizon
- More importantly, measuring the width of a street while not also moving towards the other side is introducing more imprecision, because the more you tilt your phone to the horizon, the bigger the distance covered per degree. I measured a 9m wide road, but the measurement was 8m. That's ~10% imprecision or 3x the imprecision as measured yesterday. When I crossed to the other side, ARCore corrected this and one could see clearly that the line stopped short of the curb.

So, especially because of the second point, this is not a good solution. I'll implement a message that suggest the user should cross the road / keep near the arrow.

**Comment by hb0nd** [comment]
This is a _really_ cool feature! The only improvements i'd like are:

- Making the arrow a bit smaller (or using some sort of crosshair instead) 
- Making the measurement line a bit transparent, so you can see what you're doing more easily.

I also wonder what else it can be used for, for instance measuring building heights, fence heights etc
